### ISSUE TITLE
High Q-values Causing NaN Policy in PriorDescent

### ISSUE THREAD

**Issue by Minghui-Wu** [radar-research-lab/MFGLib#41]
**Issue:** When running PriorDescent on a long-horizon MDP, the Q-values become extremely high, leading to overflow in torch.exp(Q) and resulting in a NaN policy.

**Steps to Reproduce:**
```
instance = Environment.beach_bar(n=10, bar_loc=2, T=15)
solns, expls, runtimes = PriorDescent(eta=1, n_inner=1, update_initial=False).solve(instance, max_iter=500, verbose=True)
```

**Comment by junziz** [comment]
Thanks for flagging this. Such an overflow is expected in general, especially due to the exponential involved here. In these cases it's just an indication of needing to adjust the hyperparameters for PriorDescent, especially increasing `eta` (which will down-weight the Q-values in the `torch.exp`). Intuitively, larger `eta` means smaller step-sizes/learning rates, which is needed for problems with larger scales in general (like the relatively long horizon 15 here). For example, the following would resolve your problem and get to the default 1e-3 tolerance within 500 iterations (of course, the exact performance and good hyperparameters choices can be different on your computer due to potential system differences). You can also try different `n_inner`, and sometimes larger `n_inner` (like 10) can work better. Lastly, you can also use the `.tune` method to automatically run hyperparameters optimization for you. 
```python
solns, expls, runtimes = PriorDescent(eta=10, n_inner=1).solve(instance, max_iter=500, verbose=True)
```

Please let us know if this solves your problem. But regardless, I think we should indeed try to reimplement the weighted softmax update in PriorDescent using torch softmax instead of torch exp, as the former would probably be much more numerically stable. Will update the implementation if we find that to work better. Thanks again!

**Comment by Minghui-Wu** [comment]
Sounds great and thanks!
